torch/lib/c10d/Utils.hpp

```diff
@@ -3,18 +3,124 @@
 #include <sys/socket.h>
 #include <sys/types.h>
 
-#include <cstdlib>
+#include <chrono>
 #include <cstdint>
+#include <cstdlib>
 #include <functional>
 #include <limits>
 #include <string>
 #include <system_error>
 #include <tuple>
 #include <vector>
-#include <chrono>
+
+#include <ATen/ATen.h>
+
+#include "Types.hpp"
+
+inline void hash_combine(std::size_t& seed) {}
+
+template <typename T, typename... Rest>
+inline void hash_combine(std::size_t& seed, const T& v, Rest... rest) {
+  std::hash<T> hasher;
+  seed ^= hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+  hash_combine(seed, rest...);
+}
+
+#define MAKE_HASHABLE(type, ...)                  \
+  namespace std {                                 \
+  template <>                                     \
+  struct hash<type> {                             \
+    std::size_t operator()(const type& t) const { \
+      std::size_t ret = 0;                        \
+      hash_combine(ret, __VA_ARGS__);             \
+      return ret;                                 \
+    }                                             \
+  };                                              \
+  }
+
+// Make std::vector<T> hashable
+namespace std {
+template <typename T>
+struct hash<vector<T>> {
+  std::size_t operator()(const vector<T>& v) const {
+    std::size_t ret = v.size();
+    for (const auto& e : v) {
+      hash_combine(ret, e);
+    }
+    return ret;
+  }
+};
+} // namespace std
```

Comment:
Hmm if we do this, then we should really make sure this header *never* gets included by the users. I'd consider adding `hash` instances for non-internal types a bad practice for a library (same applies to `ArrayRef` below).

I've actually written a helper for hashing that uses a separate mechanism from `std::hash`, to make it easier to support things like vectors, and making structs hashable easier (it will look for a `.hash()` method on them by default, and only then fall back to `std::hash`). It also supports multiple args, so `torch::hash(x, y, z, w)` is a valid expr. cc @ezyang as this might be useful for c10 utility discussions at some point.

Reply:
Agree. I'm fine using something else as long as long as it works with std containers.

Reply:
Let's shelve this until we figure out what to do.

Reply:
Unshelving, I'll just refer to the header you linked.